[compiler-rt] Add CMake option to enable execute-only code generation on AArch64

Il-Capitano opened the pull request:
For a full toolchain supporting execute-only code generation the runtime libraries also need to be pre-compiled with it enabled. The generic `RUNTIMES_EXECUTE_ONLY_CODE` CMake option can now be used during build configuration to enable execute-only code generation in compiler-rt.

The build option can only be enabled for a runtimes build of compiler-rt, because a recent version of Clang is needed to correctly compile assembly files with execute-only code support.

Related RFC: https://discourse.llvm.org/t/rfc-execute-only-code-support-for-runtime-libraries-on-aarch64/86180

Il-Capitano committed:
[compiler-rt] Add CMake option to enable execute-only code generation on AArch64

For a full toolchain supporting execute-only code generation the runtime
libraries also need to be pre-compiled with it enabled. For compiler-rt
this can now be enabled with the `COMPILER_RT_EXECUTE_ONLY_CODE` CMake
option during build configuration.

The build option can only be enabled for a runtimes build of
compiler-rt, because a recent version of Clang is needed to correctly
compile assembly files with execute-only code support.

Related RFC: https://discourse.llvm.org/t/rfc-execute-only-code-support-for-runtime-libraries-on-aarch64/86180

compnerd reviewed (commented):
I don't think that this properly handles MachO. Additionally, I don't understand the need for the `unique` flag on the section, nor the section id (`0`).

Il-Capitano commented:
> I don't think that this properly handles MachO.

Can you expand on this more? MachO doesn't have execute-only code AFAIK. This is intended for AArch64 ELF targets only.

> Additionally, I don't understand the need for the `unique` flag on the section, nor the section id (`0`).

`unique,0` is needed because the assembler creates an implicit `.text` section with default flags. If we tried to just do `.section .text,"axy"`, we'd get the following error: `error: changed section flags for .text, expected: 0x6`. So the execute-only `.text` section needs to be different from the implicit `.text` section. The id `0` is arbitrary, but we need to provide something anyways.

This also matches what LLVM outputs when it generates execute-only code: https://godbolt.org/z/a8xa7qsxG

compnerd commented:
> > I don't think that this properly handles MachO.
> 
> Can you expand on this more? MachO doesn't have execute-only code AFAIK. This is intended for AArch64 ELF targets only.

These macros are not ELF specific is the issue. You are changing macros that are shared across all architectures and file formats. If you could scope this down somehow to not touch these global macros, that might be easier.

> > Additionally, I don't understand the need for the `unique` flag on the section, nor the section id (`0`).
> 
> `unique,0` is needed because the assembler creates an implicit `.text` section with default flags. If we tried to just do `.section .text,"axy"`, we'd get the following error: `error: changed section flags for .text, expected: 0x6`. So the execute-only `.text` section needs to be different from the implicit `.text` section. The id `0` is arbitrary, but we need to provide something anyways.
> 
> This also matches what LLVM outputs when it generates execute-only code: https://godbolt.org/z/a8xa7qsxG

Ah, I see. I think that deserves a comment explaining why we need those flags are being added. I do wonder if we should simply drop the `unique,0` and instead build the builtins as `-ffunction-sections` instead as that places the text into a separate section.

Il-Capitano commented:
> > > I don't think that this properly handles MachO.
> > 
> > 
> > Can you expand on this more? MachO doesn't have execute-only code AFAIK. This is intended for AArch64 ELF targets only.
> 
> These macros are not ELF specific is the issue. You are changing macros that are shared across all architectures and file formats. If you could scope this down somehow to not touch these global macros, that might be easier.

The change is adding a `.text` or `.section .text,"axy",unique,0` directive before the function definitions, which just makes sure we're in the `.text` section. The execute-only case is guarded by `defined(__aarch64__) && defined(__ELF__)`, so shouldn't affect any other targets. The alternative would be to change every AArch64 assembly file individually, but I don't think that would be better.

The cases where functions are put in differently named sections don't seem to use these macros anyways.

> > > Additionally, I don't understand the need for the `unique` flag on the section, nor the section id (`0`).
> > 
> > 
> > `unique,0` is needed because the assembler creates an implicit `.text` section with default flags. If we tried to just do `.section .text,"axy"`, we'd get the following error: `error: changed section flags for .text, expected: 0x6`. So the execute-only `.text` section needs to be different from the implicit `.text` section. The id `0` is arbitrary, but we need to provide something anyways.
> > This also matches what LLVM outputs when it generates execute-only code: https://godbolt.org/z/a8xa7qsxG
> 
> Ah, I see. I think that deserves a comment explaining why we need those flags are being added. I do wonder if we should simply drop the `unique,0` and instead build the builtins as `-ffunction-sections` instead as that places the text into a separate section.

You're right, I'll add a comment clarifying this. For dropping `unique,0` do you mean to add `.section .text.name` in the function definition macros? That could work to get rid of the `unique,0` suffix.

compnerd commented:
> > > > I don't think that this properly handles MachO.
> > > 
> > > 
> > > Can you expand on this more? MachO doesn't have execute-only code AFAIK. This is intended for AArch64 ELF targets only.
> > 
> > 
> > These macros are not ELF specific is the issue. You are changing macros that are shared across all architectures and file formats. If you could scope this down somehow to not touch these global macros, that might be easier.
> 
> The change is adding a `.text` or `.section .text,"axy",unique,0` directive before the function definitions, which just makes sure we're in the `.text` section. The execute-only case is guarded by `defined(__aarch64__) && defined(__ELF__)`, so shouldn't affect any other targets. The alternative would be to change every AArch64 assembly file individually, but I don't think that would be better.

Part of the change does impact a place where there was no use of the macro and we placed text into `.section __TEXT,__text` which is the proper spelling.

> The cases where functions are put in differently named sections don't seem to use these macros anyways.
> 
> > > > Additionally, I don't understand the need for the `unique` flag on the section, nor the section id (`0`).
> > > 
> > > 
> > > `unique,0` is needed because the assembler creates an implicit `.text` section with default flags. If we tried to just do `.section .text,"axy"`, we'd get the following error: `error: changed section flags for .text, expected: 0x6`. So the execute-only `.text` section needs to be different from the implicit `.text` section. The id `0` is arbitrary, but we need to provide something anyways.
> > > This also matches what LLVM outputs when it generates execute-only code: https://godbolt.org/z/a8xa7qsxG
> > 
> > 
> > Ah, I see. I think that deserves a comment explaining why we need those flags are being added. I do wonder if we should simply drop the `unique,0` and instead build the builtins as `-ffunction-sections` instead as that places the text into a separate section.
> 
> You're right, I'll add a comment clarifying this. For dropping `unique,0` do you mean to add `.section .text.name` in the function definition macros? That could work to get rid of the `unique,0` suffix.

Right, that is what I was thinking. For static linking, that might also enable DCE, which can be a benefit.

Il-Capitano commented:
> Part of the change does impact a place where there was no use of the macro and we placed text into `.section __TEXT,__text` which is the proper spelling.

Is there a meaningful difference between `.text` and `.section __TEXT,__text`? From what I found, and my local testing, the former is an alias of the latter. There are instances in compiler-rt, where `.text` is used on MachO as well (e.g. `compiler-rt/lib/orc/sysv_reenter.arm64.S`).

> Right, that is what I was thinking. For static linking, that might also enable DCE, which can be a benefit.

This is an orthogonal change to mine though. Would you be fine with merging this PR as is, and you/someone else preparing a separate patch for that? Alternatively that change could be done first, and I'd be happy to do a rebase on this PR accommodating those changes.

Il-Capitano committed:
Add comment describing why 'unique,0' is needed

arichardson commented on compiler-rt/lib/orc/elfnix_tls.aarch64.S:
```diff
@@ -15,7 +15,11 @@
 
 #define REGISTER_SAVE_SPACE_SIZE     32 * 24
 
+#if defined(COMPILER_RT_EXECUTE_ONLY_CODE)
```
Why can;t we include the assembly.h header file here and use TEXT_SECTION?

Il-Capitano replied:
Good point, I think I was being too conservative with adding extra includes. Fixed now in both files.

compnerd commented:
> > Part of the change does impact a place where there was no use of the macro and we placed text into `.section __TEXT,__text` which is the proper spelling.
> 
> Is there a meaningful difference between `.text` and `.section __TEXT,__text`? From what I found, and my local testing, the former is an alias of the latter. There are instances in compiler-rt, where `.text` is used on MachO as well (e.g. `compiler-rt/lib/orc/sysv_reenter.arm64.S`).

Oh, if `.text` is treated as an alias, I think that should be fine.

> > Right, that is what I was thinking. For static linking, that might also enable DCE, which can be a benefit.
> 
> This is an orthogonal change to mine though. Would you be fine with merging this PR as is, and you/someone else preparing a separate patch for that? Alternatively that change could be done first, and I'd be happy to do a rebase on this PR accommodating those changes.

Well, it is only coming up due to the changes you are making. I am thinking about how to simplify and make the code easier to maintain.

Il-Capitano committed:
Use TEXT_SECTION macro in orc

Il-Capitano commented:
@ldionne [commented on my RFC](https://discourse.llvm.org/t/rfc-execute-only-code-support-for-runtime-libraries-on-aarch64/86180/4) about using a single CMake option for all runtime libraries.

If there aren't any opposing views, I'm fine with making that change, so if you have any thoughts (either in support of the change or opposing it), please comment on the RFC. Thanks!

Il-Capitano committed:
Simplify CMake configuration based on RFC feedback

Il-Capitano commented on compiler-rt/CMakeLists.txt:
```diff
@@ -307,6 +307,11 @@ option(COMPILER_RT_USE_BUILTINS_LIBRARY
 
+if (LLVM_EXECUTE_ONLY_CODE AND NOT LLVM_RUNTIMES_BUILD)
+  message(SEND_ERROR "LLVM_EXECUTE_ONLY_CODE is only supported "
+                     "for runtimes build of compiler-rt.")
```
I'll remove it then, thanks.

Il-Capitano committed:
Don't check if we're doing a runtimes build

A non-runtimes build of compiler-rt is deprecated and should be removed
in the next release entirely, so we don't have to check for it.

Il-Capitano committed:
Rename LLVM_EXECUTE_ONLY_CODE to RUNTIMES_EXECUTE_ONLY_CODE

Il-Capitano commented:
Do you have any more thoughts on this PR?

I would like to merge this by the end of the month, before the next release period starts.
